File: JavaScript/selectionSort.js
/**
 * Sort array using selection sort algorithm
 * @param {Array} arr
 */

function selectionSort(arr){
  let index = 0;
  let sorted = false;

  while (index < arr.length && !sorted){
    let iSmallest = index;
    sorted = true;
    
    // use for loop instead of forEach to avoid
    // looping over already sorted items
    for (let i = index + 1; i < arr.length; i++){
      if (arr[i] <= arr[iSmallest]){
        sorted = false;
        iSmallest = i;
      }
    }

    let [ smallestItem ] = arr.splice(iSmallest, 1);
    arr.splice(index, 0, smallestItem);

    index++;
  }

  return arr;
}


console.log(selectionSort([2, 3, 5, 1, 7, 2, 8, 9]))

File: Python/selectionSort.py
def selectionSort(l: list) -> list:
  """
  consumes a list and sorts using the selection sort algorithm
  returns the sorted list
  >>> selectionSort([7, 1, 3, 5, 4, 6, 2]) -> [1, 2, 3, 4, 5, 6, 7]
  """
  i = 0

  while len(l) > 0:
    smallest = l[0]
    for item in l:
      if item < smallest:
        smallest = item
      
    l.remove(smallest)
    l.insert(i, smallest)
    i += 1
    
  return l

